### PR TITLE
Do as much gradients at once as possible in trainer

### PR DESCRIPTION
**Pull request by jlibovicky**
The trainer was probably more memory demanding because it was computing the gradients of the loss and regularization separately and then summing the up.

@tomkocmi, please chceck how the memory requirements of your experiments change in with this branch.

### CODE REVIEW

File: neuralmonkey/trainers/generic_trainer.py [hunk 1]
```diff
@@ -41,17 +41,21 @@ def __init__(self, objectives: List[Objective],
...
-        gradients = _sum_gradients(partial_gradients)
+        # if the objective does not have its own gradients,
+        # just use TF do the derivative
+        objectives_without_gradients = sum(
+            o.loss for o in objectives
+            if  o.gradients is None) + l1_cost + l2_cost
+        implicit_gradients = self._get_gradients(objectives_without_gradients)
+
+        # objectives that have their gradients explictly computed
+        other_gradients = [
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: tohohle jsem si před tím nevšimnul - to je nějaká příprava na mixér nebo minimum risk training?

File: neuralmonkey/trainers/generic_trainer.py [hunk 2]
```diff
@@ -41,17 +41,21 @@ def __init__(self, objectives: List[Objective],
...
-                partial_gradients.append(gradients)
-            else:
-                partial_gradients.append(objective.gradients)
-        partial_gradients += [self._get_gradients(l)
-                              for l in [l1_cost, l2_cost] if l != 0.]
-
-        gradients = _sum_gradients(partial_gradients)
+        # if the objective does not have its own gradients,
+        # just use TF do the derivative
+        objectives_without_gradients = sum(
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: nazev ti vnucuje premejslet jako o listu, ale je to skalar.. nazval bych to spis differentiable_loss nebo tak nejak

File: neuralmonkey/trainers/generic_trainer.py [hunk 3]
```diff
@@ -41,17 +41,21 @@ def __init__(self, objectives: List[Objective],
...
-                gradients = self._get_gradients(objective.loss)
-                partial_gradients.append(gradients)
-            else:
-                partial_gradients.append(objective.gradients)
-        partial_gradients += [self._get_gradients(l)
-                              for l in [l1_cost, l2_cost] if l != 0.]
-
-        gradients = _sum_gradients(partial_gradients)
+        # if the objective does not have its own gradients,
+        # just use TF do the derivative
```
Comment (jindrahelcl) [inline neuralmonkey/trainers/generic_trainer.py]: ... TF **to** do ...

File: neuralmonkey/trainers/generic_trainer.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Přesně tak.

**Pull request merged**